plasmapy/tests/helpers/exceptions.py

```diff
@@ -0,0 +1,105 @@
+"""Exceptions that describe different types of test failures."""
+
+__all__ = [
+    "ExceptionMismatchFail",
+    "InvalidTestError",
+    "MissingExceptionFail",
+    "MissingWarningFail",
+    "TestFailed",
+    "TypeMismatchFail",
+    "UnexpectedExceptionFail",
+    "UnexpectedResultFail",
+    "UnexpectedWarningFail",
+    "WarningMismatchFail",
+]
+
+import pytest
+
+
+class TestFailed(pytest.fail.Exception):
+    """
+    Base exception for test failures.
+
+    Notes
+    -----
+    This exception was derived from `~pytest.fail.Exception`, which in
+    turn was derived from `BaseException` (not `Exception`).
+    """
+
+    pass
+
+
+class TypeMismatchFail(TestFailed):
+    """
+    Exception for when the type of the actual result differs from the
+    type of the expected result.
+    """
+
+    pass
+
+
+class MissingExceptionFail(TestFailed):
+    """
+    Exception for when the expected exception is not raised.
+    """
+
+    pass
+
+
+class UnexpectedExceptionFail(TestFailed):
+    """
+    Exception for when an exception is raised unexpectedly.
+    """
+
+    pass
+
+
+class ExceptionMismatchFail(UnexpectedExceptionFail, MissingExceptionFail):
+    """
+    Exception for when an exception is expected, but a different
+    exception is raised.
+    """
+
+    pass
+
+
+class UnexpectedWarningFail(TestFailed):
+    """
+    Exception for when a warning is issued unexpectedly.
+    """
+
+    pass
+
+
+class MissingWarningFail(TestFailed):
+    """
+    Exception for when an expected warning is not issued.
+    """
+
+    pass
+
+
+class WarningMismatchFail(UnexpectedWarningFail, MissingWarningFail):
+    """
+    Exception for when a warning is expected, but one or more
+    different warnings were issued instead.
+    """
+
+    pass
+
+
+class UnexpectedResultFail(TestFailed):
+    """
+    Exception for when the actual result differs from the expected
+    result by more than the allowed tolerance.
+    """
```

Comment:
Might want to generalize the wording a little more, since I can imagine an `UnexpectedResultFail` that is not involve `isclose` or `allclose`.  For example, arrays of differing sizes.